test/framework/k8s_cluster.go

```diff
@@ -0,0 +1,310 @@
+		},
+		1, defaultRetries, defaultTiemout)
+
+	kumacp_pods := c.GetKumaCPPods()
+	Expect(len(kumacp_pods)).To(Equal(1))
+
+	k8s.WaitUntilPodAvailable(c.t,
+		c.GetKubectlOptions(kumaNamespace),
+		kumacp_pods[0].Name,
+		10, 3*time.Second)
+
+	c.localCPPort = c.PortForwardKumaCP()
+
+	err = c.kumactl.KumactlConfigControlPlanesAdd(c.name, "http://localhost:"+strconv.FormatUint(uint64(c.localCPPort), 10))
+	Expect(err).ToNot(HaveOccurred())
+
+	return nil
+}
+
+func (c *K8sCluster) DeleteKuma() error {
+	c.CleanupPortForwards()
+
+	yaml, err := c.kumactl.KumactlInstallCP()
+	if err != nil {
+		return err
+	}
+
+	err = k8s.KubectlDeleteFromStringE(c.t,
+		c.GetKubectlOptions(),
+		yaml)
+
+	c.WaitKumaNamespaceDelete()
+
+	return err
+}
+
+func (c *K8sCluster) DeleteKumaNamespace() error {
+	return k8s.DeleteNamespaceE(c.t,
+		c.GetKubectlOptions(),
+		kumaNamespace)
+}
+
+func (c *K8sCluster) WaitKumaNamespaceDelete() {
+	retry.DoWithRetry(c.t, "Wait the Kuma Namespace to terminate.", 2*defaultRetries, defaultTiemout,
+		func() (string, error) {
+			_, err := k8s.GetNamespaceE(c.t,
+				c.GetKubectlOptions(),
+				kumaNamespace)
+			if err != nil {
+				return "", nil
+			}
+			return "", fmt.Errorf("Kuma Namespace still not terminated.")
+		})
+}
+
+func (c *K8sCluster) PortForwardKumaCP() uint32 {
+	pods := c.GetKumaCPPods()
+	if len(pods) < 1 {
+		fmt.Println("No kuma-cp pods found for port-forward.")
+		return 0
+	}
+	name := pods[0].Name
+
+	//find free local port
+	localPort, err := util_net.PickTCPPort("127.0.0.1", 32000+kumaCPAPIPort, 42000+kumaCPAPIPort)
+	if err != nil {
+		fmt.Println("No free port found in range: ", 32000+kumaCPAPIPort, " - ", 42000+kumaCPAPIPort)
+		return 0
+	}
+
+	c.PortForwardPod(kumaNamespace, name, localPort, kumaCPAPIPort)
+	return localPort
+}
+
+func (c *K8sCluster) PortForwardPod(namespace, name string, localPort, remotePort uint32) {
+	config, err := clientcmd.BuildConfigFromFlags("", c.kubeconfig)
+	if err != nil {
+		fmt.Println("Error building config")
+		return
+	}
+	roundTripper, upgrader, err := spdy.RoundTripperFor(config)
+	if err != nil {
+		panic(err)
+	}
+
+	path := fmt.Sprintf("/api/v1/namespaces/%s/pods/%s/portforward", namespace, name)
+	hostIP := strings.TrimLeft(config.Host, "htps:/")
+	serverURL := url.URL{Scheme: "https", Path: path, Host: hostIP}
+
+	stopChan, readyChan := make(chan struct{}, 1), make(chan struct{}, 1)
+	out, errOut := new(bytes.Buffer), new(bytes.Buffer)
+
+	dialer := spdy.NewDialer(upgrader, &http.Client{Transport: roundTripper}, http.MethodPost, &serverURL)
+
+	ports := []string{strconv.FormatUint(uint64(localPort), 10) + ":" + strconv.FormatUint(uint64(remotePort), 10)}
+
+	forwarder, err := portforward.New(dialer, ports, stopChan, readyChan, out, errOut)
+	if err != nil {
+		panic(err)
+	}
+
+	go func() {
+		for range readyChan { // Kubernetes will close this channel when it has something to tell us.
+		}
+		if len(errOut.String()) != 0 {
```

Comment:
That looks like concurrent access to `errOut`, is it ok?

Reply:
Good catch. Let me think about it.

Reply:
Actually no, once the channel is closed, the errOut won't be used anymore, so it should be safe to use it here.